pepper-apis/src/main/java/org/broadinstitute/ddp/db/ActivityDefStore.java

```diff
@@ -161,6 +182,37 @@ public void setActivityDef(String studyGuid, String activityCode, String version
         }
     }
 
+    @VisibleForTesting
+    public void clearCachedActivityData() {
+        synchronized (lockVar) {
+            activityDtoMap.clear();
+            activityDefMap.clear();
+        }
+    }
+
+    public String findValidationRuleMessage(
+            Handle handle, RuleType ruleType, Long hintTemplateId, long langCodeId, long timestamp) {
+        synchronized (lockVar) {
+            return validationRuleMessageMap.computeIfAbsent(ruleType.name() + langCodeId, message ->
+                    detectValidationRuleMessage(handle, ruleType, hintTemplateId, langCodeId, timestamp));
+        }
+    }
+
+    private String detectValidationRuleMessage(
+            Handle handle, RuleType ruleType, Long hintTemplateId, long langCodeId, long timestamp) {
+        String correctionHint = null;
+        if (hintTemplateId != null) {
+            correctionHint = i18nContentRenderer.renderContent(handle, hintTemplateId, langCodeId, timestamp);
+        }
+        if (correctionHint != null) {
+            return correctionHint;
```

Comment:
I don't think `correctionHint` belongs here. ActivityDefStore should only deal with caching the default validation message that corresponds to each type of validation rule.